Feature/authentication

keving17 opened the pull request:
This pull request fixes the authentication feature that was pushed earlier. CSS formatting and configuration is fixed so the login portal is functional. When someone logs in, they should be able to access data directly from the online database. For now, I've limited this to only pull data with the location "localhost:3000/teachermoments/turner?KevinTesting20180319"

Accessing my test data
<img width="1427" alt="screen shot 2018-03-19 at 7 18 47 pm" src="https://user-images.githubusercontent.com/11862658/37627473-f36c7ba0-2baa-11e8-9c5b-57fbabc2900e.png">

No access to other data
<img width="840" alt="screen shot 2018-03-19 at 7 38 40 pm" src="https://user-images.githubusercontent.com/11862658/37627923-2868b830-2bad-11e8-8fb0-501f33c8a75a.png">

Something that I am unsure about is how to get the consent-latest.json file into the system.

keving17 commented:
Also, I've yet to test if this system is able to grab the wav files from s3. Any thoughts on how to do that without deploying?

keving17 commented on client/src/EmailLinkLoginPage.js:
```diff
@@ -1,8 +1,7 @@
-import Interactions from './Interactions.js';
-// import TeacherMoments from './TeacherMoments.js';
+import './LoginPage.css';
+import TeacherMoments from './TeacherMoments.js';
```
makes sense. done!

keving17 commented on client/src/Analyses.js:
```diff
@@ -1,393 +1,430 @@
-// import Filters from './Filters.js';
```
Hmmm. It seems that github had some trouble figuring out what changed in the Analyses.js file. I'm still using the Filter.js code - the new Analyses.js still has this same import on line 1.

As for cutting a new branch - yeah that makes sense. I think i was mentally grouping all the work around getting the self serve portal up and running under the feature/authentication branch. I'll be sure to cut new branches moving forward for each individual feature

kevinrobinson replied:
ah so it was all commented out before and this fixes that, got it!

keving17 commented on server/domain.js:
```diff
@@ -1,7 +1,7 @@
   getDomain: function(request) {
     return (process.env.NODE_ENV === 'development')
-      ? 'http://localhost:5000'
+      ? 'http://localhost:3000'
```
Hmmmm so how I understand this is that previously we worked on localhost:5000. Using the same code, I was running into a problem where the links I was generating were not directing to localhost:3000 where teacher moments was being hosted locally. With the new build, the dev server is at port 4000 and the client that is hot-loaded is at port 3000. I really only use this getDomain function to create urls in development mode that direct us to the client hosted at port 3000.

Are you saying to add a comment here to explain this bit? I'm not sure I understand this comment too well.

kevinrobinson replied:
Yeah, I think we're saying the same thing, and that adding a comment would be good - this is pointing to a port that's *not* the same port this actual server is running at or the same as `config.port` (which is what I think folks would expect).

keving17 replied:
done!

keving17 commented on client/src/EmailLinkLoginPage.js:
```diff
@@ -82,13 +84,17 @@ class EmailLinkLoginPage extends Component {
         return (
-          <Interactions email={email.toLowerCase()} token={token}/>
-          // <TeacherMoments email={email.toLowerCase()} token={token}/>
+          <TeacherMoments email={email.toLowerCase()} token={token}/>
```
renamed!

keving17 commented on client/src/EmailLinkLoginPage.js:
```diff
@@ -25,6 +24,14 @@ class EmailLinkLoginPage extends Component {
+
+  componentWillUnmount() {
+    document.body.style.backgroundColor = null;
+  }
```
So my intuition with changing the background colors was to change it in the css file. However, it seems that setting the background of the outermost div in the render method just gives us a white box on a black background.
<img width="720" alt="screen shot 2018-03-27 at 8 56 26 pm" src="https://user-images.githubusercontent.com/11862658/38002631-6cc34c7a-3201-11e8-9300-574453eea56d.png">

I get the right behavior if I move the 
```
document.body.style.backgroundColor = "white"
```
to the render method but I'm not sure if that solves the problem. My understanding is that you're saying my fix might not work for mobile? If I do keep my workaround, how do I factor out the color change into a new component and add it back in? I'm not quite following your explanation.

kevinrobinson replied:
Got it!  Yeah, looking at CSS is a good intuition as to where to start.  There's some complexity here in that the CSS is written directly into the HTML (see link above) to always apply to the body tag, since the app was initially setup to render in the full screen on mobile, but to be styled to look the size of a mobile app if you're on desktop.  That's why there's no obvious way to change that.

So to change that code, you'll have to test across mobile and desktop screen sizes and make sure things still work.  With this work, it'd be great to avoid that so your workaround strategy is great. 👍 

My suggestion here is that you add a comment about this, and since it's kind of a weird quirk and you've had to add it in multiple places here already (and will for other pages that you want to have all white background), then to factor it into it's only tiny component that just sets this style and explains why it's doing that in a comment.  Alternately, you could work around and keep the black background styling (that's what [this page](https://teachermoments.teachingsystemslab.org/equity) does).

keving17 replied:
gotcha. How would I go about factoring this styling quirk into its own separate component? Should I just make a new js file in the research folder, import the component in LoginPage.js and EmailLinkLoginPage.js, and render the component? I think I'm a bit lost on how this happens with the React components

kevinrobinson replied:
Ah cool this is a good time to learn then!  :)  Yep, it's just like all the other components you've been working with and writing, you just have to make the new file and import like you described.

keving17 replied:
done!

keving17 commented on server/index.js:
```diff
@@ -10,18 +10,36 @@ var ApplesEndpoint = require('./endpoints/apples.js');
-const s3 = createS3Client();
-const mailgunEnv = createMailgunEnv();
+// const s3 = createS3Client();
+// const mailgunEnv = createMailgunEnv();
```
oh. these should be deleted. Similar to how Swipe Right declares the mailgunEnv in the config, I create the s3 client and the mailgun env inside the config. 

I noticed this caused a problem when calls like `ReviewLoginEndpoint.createReview({queryDatabase, mailgunEnv})` used the mailgunEnv within the curly brackets. I didn't know why we used curly brackets and the code seemed to work fine without it so I changed the calls to not use the brackets. A similar problem appeared with the s3 variable as well.

kevinrobinson replied:
ah yeah got it, moving to config is great.  The curly brackets aren't magic - I commented on this elsewhere, the shorthand of `{kevin}` converts to `{kevin: kevin}` so you can only use it with a variable name that's in scope.  You can't use it with `{config.kevin}`but you can just do the normal object syntax then like `{kevin: config.kevin}`

keving17 replied:
yep! done!

keving17 commented on server/index.js:
```diff
@@ -170,18 +188,42 @@ app.get('/server/questions', questionAuthoringAuth, function(request, response){
+  //   console.log('got request')
+  //   dataEndpoint.bind(null, pool);
+  //   console.log('after func')
+  // });
```
cleaned up!

keving17 commented on server/index.js:
```diff
@@ -170,18 +188,42 @@ app.get('/server/questions', questionAuthoringAuth, function(request, response){
 
 
 // Related to the read path for reviewing responses, and for fetching audio files
-app.post('/server/reviews/create', ReviewLoginEndpoint.createReview({queryDatabase, mailgunEnv}));
```
ahh yes, I talked a bit about this with relation to the mailgunEnv and s3 variables. After moving those within the config, these calls broke when I used config.mailgunEnv instead of mailgunEnv. I don't really understand what restrictions the curly brackets enforce but I noticed that by removing the curly brackets, things worked fine again. I'm still not sure why this happens so if we feel strongly about having the curly brackets, I can look into this more?

kevinrobinson replied:
Got it.  So the curly brackets aren't doing anything magical, they're either creating an object, or pulling values out of an object.

```
ReviewLoginEndpoint.createReview({queryDatabase, mailgunEnv})

// is the same as:
const object = {
  queryDatabase: queryDatabase,
  mailgunEnv: mailgunEnv
};
ReviewLoginEndpoint.createReview(object);
```

So that passes one argument to the function, an object with two keys.

In contrast:
```
ReviewLoginEndpoint.createReview(queryDatabase, mailgunEnv)
```
calls the function with two arguments.

I don't know the details of what you changed here, but what might have happened before is that the shorcut `{foo}` expands to `{foo: foo}`.  But you can't do `{config.foo}` since it doesn't make sense to convert that to a key and value as`{config.foo: config.foo}`.  In that case you can just say `{foo: config.foo}`.

keving17 replied:
This makes sense now! I've fixed this back to using the curly brackets.

keving17 commented on server/database.js:
```diff
@@ -0,0 +1,39 @@
+  // const test_location = request.headers['x-teachermoments-location'];
+  // console.log(test_location)
+
+  const location = 'http://localhost:3000/teachermoments/turner?KevinTesting20180319';
```
done! I actually hadn't thought of this. Thanks!

keving17 commented on server/database.js:
```diff
@@ -0,0 +1,39 @@
+      }
+      else{
+        console.log('Could not find data with url:', location);
+        return response.status(500).end();
```
that makes sense. Nothing broke so it shouldn't be a 500. done.

keving17 commented on server/database.js:
```diff
@@ -0,0 +1,39 @@
+function dataEndpoint(pool, request, response) {
```
done!

kevinrobinson replied:
maybe you did this locally?

keving17 replied:
hmmmm I added a commit last night at https://github.com/mit-teaching-systems-lab/threeflows/commit/62f24828d586a2f8a89042e64d8f4dbd97794aa3

kevinrobinson replied:
ah sorry I think I mis-githubbed

keving17 commented on server/database.js:
```diff
@@ -0,0 +1,39 @@
+    .then(results => {
+      if (results.rowCount >= 1){
+        console.log('found data');
+        console.log(results);
```
removed

keving17 commented on server/database.js:
```diff
@@ -0,0 +1,39 @@
+function dataEndpoint(pool, request, response) {
+  // const test_location = request.headers['x-teachermoments-location'];
+  // console.log(test_location)
```
I left this here because I wasn't sure how to go about deciding what url to look for. For now, I hard code this function to use 
```
const location = `${domain}/teachermoments/turner?KevinTesting20180319`;
```

but in production, people should be able to query for any location. Is the idea to make a middleman function to first confirm that the current user has access to data with that url?

keving17 replied:
I'm removing this code for now but I do think the location variable will end up coming from the header of the request

kevinrobinson replied:
Okay, so I commented over https://github.com/mit-teaching-systems-lab/threeflows/pull/348#discussion_r177618684, let's talk more over there :)

The main idea is that you can't trust the `location` param for the client blindly, and somewhere on the server has to check this (or just hardcode for now and build that second step in the next PR).

keving17 replied:
got it!

keving17 commented on client/src/TeacherMoments.js:
```diff
@@ -0,0 +1,431 @@
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+        'x-teachermoments-location': location,
```
oh this got hidden since I renamed TeacherMoments.js

This makes sense. It sounds like a need a function that queries the token against a table that says who is authorized to see what data. I'll work on that!

keving17 replied:
Actually, when you say "The server code is deciding what the user can see and limiting it to just that playtest", you're referring to me hard coding the location variable in database.js right?

I'm thinking I will still need the location in the header to communicate what data to query for in databse.js. I think I would need to use the token to confirm that this user can query for data at a specific location. That could probably be something like:
```
checkAccess(token, location)
  .then ( fetch with just the location header)
  .catch( console.log("you are not authorized to access this data"));
```
And so yeah, I shouldnt need the email once I have the token

kevinrobinson replied:
Cool, this is a great explanation, I think I'm following this.

So to show the user data, one thing the server needs to know is: "what set of data should I should them?"  And there's two parts: 1) what dataset the user wants, and 2) what they should be authorized to see.

When you said:
> It sounds like a need a function that queries the token against a table that says who is authorized to see what data. I'll work on that!
That's a great approach to this!  And I think what you'll want to do for sure soon.

But I was suggesting that for part 1) you just hard code that on the server so the only thing users can ask for is data for `?KevinTesting20180319` or whatever.  And then in this pull request all you need to do is 2) lookup to see if the user's token is valid and they have researcher access.  The downside to this is that in the short term all researchers can only access that same set of data you hardcoded.

Your suggestion is better overall for sure, so my comment here is just a suggestion that maybe you ship the hardcoded "what dataset you want" for now, and all you have to do is do the authorization check.  If you want to do both at once, and that seems better or simpler for any reason that's great too!

keving17 replied:
gotcha! I think it makes sense to do just the hardcoded version first to finish off this pull request :)

keving17 commented on client/src/TeacherMoments.js:
```diff
@@ -0,0 +1,431 @@
+    //     {json && this.renderStats('email', statsForUsers)}
+    //     {json && this.renderStats('questionText', statsForQuestions)}
+    //   </div>
+    // );
```
deleted!

keving17 commented on server/database.js:
```diff
@@ -0,0 +1,39 @@
+
+  const location = 'http://localhost:3000/teachermoments/turner?KevinTesting20180319';
+
+  const values = [location];
```
This makes a lot of sense! I'll get started on working on things locally. Looking ahead, how exactly do I work with the production database? I don't think I've ever done that part before.

kevinrobinson replied:
Sounds good!  Here's the Heroku Postgres docs: https://devcenter.heroku.com/articles/heroku-postgresql.  Keep in mind that operations there are running on the production database, so if you run INSERT UPDATE or DELETE operations you will muck with real production data.  Running `SELECT` queries will just read and are generally much safer :)

keving17 replied:
Alright, so I just added all of the extra tables I need to the production database. I've tested that my code does run as expected in a local development database. I also have a script to transcribe all of the emails in `latest-consent.json` into a psql database which I have tested locally.

Is there any particular reason you said to manually insert the emails? Should I not use the script?
>F. Take that set of emails, and manually run SQL comments to insert them into the production database.

keving17 replied:
>G. Test and verify that works as expected.

Also, is there a way to test this works in production outside of merging the code? Do I just get rid of the NODE_ENV=development tag? That seems dangerous

kevinrobinson replied:
Awesome!  Your script is even better.

It changes the order of the steps here a bit, which is fine.  You'll have to deploy before you run the script.  So then when this is all ready, you can replace steps F and G with: a) deploy, b) verify that you can sign in as a researcher, but can't see any data because none of it is consented and then c) run your script and d) verify you can now see only consented data.

keving17 replied:
hmmmmm can I deploy without merging this work into master?

kevinrobinson replied:
You can but if you want to deploy you should merge instead.  Here I mean "merge and deploy" :)

keving17 commented on client/src/Analyses.js:
```diff
@@ -1,393 +1,430 @@
+import Filters from './Filters.js';
+import _ from 'lodash';
+import LATEST_CONSENT from './PRIVATE_SENSITIVE_DATA/consent-teacher-moments/consented-latest.json';
+import HMTCA_PAPER_CONSENT from './PRIVATE_SENSITIVE_DATA/consent-teacher-moments/consented-hmtca-paper.json';
```
commit https://github.com/mit-teaching-systems-lab/threeflows/pull/348/commits/7dc1a61ea59ebf5bb4c90a97196412c1ad73f506 resolves this. Moved consent emails to a database that we check before showing data

keving17 commented:
As an update - Looking back at your original review
> My suggestion would be:
> - [ ] address minor comments or changes in this PR
> - [ ] work on steps for consent data
> - [ ] look at loading audio files from S3 locally

- [x] address minor comments or changes in this PR
I think I've gone through and fixed all the comments you had made before.

- [ ] work on steps for consent data
I've written and tested a script that can port all the emails in latest-consent.json into a psql table. Will try this out on the production database today.

- [ ] look at loading audio files from S3 locally
I commented on this over email but I deleted my s3 credentials when I reset my repository. Can I get these from you on Monday?

kevinrobinson commented on scripts/updateConsent.js:
```diff
@@ -0,0 +1,28 @@
+
+// Create a new database for test or development mode, and 
+// create the tables for the database.
+const databaseName = process.argv[2];
```
This is great, scripting this is even better than adding this manually!  And also awesome how you put it as a command in the the `package.json` so folks can discover it.  One thing that might help is adding a line or two to the README about how to use this.

Also, if I read it right, this script expects the `consent-latest.json` format, which was produced by a separate script in the private repo that ran on the TSV file downloaded from Google Sheets.  So if I follow, I think for this to fully work you'll need to either check in a version of that file, or incorporate it into this script here.

keving17 replied:
sounds good. I've added another script that is basically a replica of the one from the private repo and added extra commands in the `package.json` to use them.

The README is currently completely out of date. I'll work on that bit after this Friday's playtest!

kevinrobinson commented on scripts/updateConsent.js:
```diff
@@ -0,0 +1,28 @@
+  const sql = `
+    INSERT INTO consented_email (email,audio,permission,consent) 
+    SELECT x,TRUE,TRUE,TRUE 
+    FROM unnest(ARRAY[${emailString}]) x;`;
```
⚠️ Making SQL commands out of strings allows a class of security vulnerabilities called "SQL injection" and is really dangerous :)

![image](https://user-images.githubusercontent.com/1056957/38222059-c0968b56-36b0-11e8-9b93-fa7c3ed99bee.png)

So here you can imagine if someone filled out our consent form with some crazy kind of email address, they might be able to get this string to be something that we wouldn't want.

If you want to input with a SQL statement, you should never form the SQL string manually, but instead use either "escaping" functions that the library provides, or a method that lets you pass in arguments separately.  Here's an example from elsewhere in this codebase that you can use as a model: https://github.com/mit-teaching-systems-lab/threeflows/blob/1696e7f01886121d9154123dfcd6819bfd051747/server/migrations/20160718-001-insert-evaluations.js#L38

keving17 replied:
Haha. I've seen that comic before!

This is my bad. I've been following a template for sql queries and this slipped my mind while I hacked this bit together. I've pushed a fix!

kevinrobinson commented on scripts/db-create.js:
```diff
@@ -58,7 +55,7 @@ function createTables(database) {
       email text,
       link text,
-      timestampz timestamptz
+      timestamp timestamptz
```
I see you had to change this in the code below before, is this broken currently because it's the wrong column name, and now your change here and in the server code fixes it?

keving17 replied:
So this part is kind of confusing. I'm not actually sure if its worth fixing. It seems that I made a typo when I first wrote this code (`timestampz` instead of `timestamptz`). I think I had originally intended to just have the name of the column match the type of data in the column (`timestamptz`).

Anyways, when I came back to work with the table this time around, it took me a while to figure out why I couldn't access this column. I'm thinking now that it might make more sense to name the column `timestamp` so that it makes sense independent from the datatype in the column. This is also because I saw some other tables use something like 
```
CREATE TABLE evidence (
      id serial primary key,
      app text,
      type text,
      version integer,
      timestamp timestamp,
      json jsonb
    );
```
which makes me question why I initially used the timestamptz datatype. I vaguely recall needing to use `timestamptz` that in order to work with timestamp values in javascript but I'm not too sure.

Bottom line is, I think it makes more sense to have the column be names `timestamp` regardless of what the datatype in the column is. I think I've fixed other parts of the code to reflect this change.

kevinrobinson replied:
Got it, thanks!  Yeah the different names can be good to communicate the different intent around whether this should be storing a timestamp in UTC or local time.

In general, we should generate and store all times as UTC, and then only at the last moment in the UI translate to local time (if we want to).  But it sounds like this might be mixed up in a few places so if you've got it working let's move on and see if it comes up again :)

kevinrobinson commented on server/database.js:
```diff
@@ -0,0 +1,65 @@
+      }
+      else{
+        console.log('Researcher is not authorized to see this data');
+        return response.status(405).end();
```
`405` is a bit different (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) `403` might be more what you want.

keving17 replied:
hmmm. Good catch. I actually haven't thought too deeply on the status codes. Fixed!

kevinrobinson commented on scripts/db-create.js:
```diff
@@ -68,7 +65,14 @@ function createTables(database) {
-      timestampz timestamptz
+      timestamp timestamptz
+    );
+    CREATE TABLE consented_email (
```
It's probably good to add `DEFAULT FALSE` to the end of these booleans too.

keving17 replied:
good idea! done!

kevinrobinson commented on server/database.js:
```diff
@@ -0,0 +1,65 @@
+
+// Spoofing the part where I confirm a token is allowed to view data from 
+// a specific location. Location is currently hard coded so this is fine.
+function checkAccess(pool, location, token) {
```
It's a good idea to explain this and your comments here are great.  But having a `checkAccess` function that doesn't actually check access can only lead to mistakes that mistakenly grants access to user data.

One thing you could do here is add in `location === `${domain}/teachermoments/turner?KevinTesting20180319` again in here.  That way you'd hard-code that the access check only allows access to one particular location.

kevinrobinson replied:
You could also move this into `getData` so that a developer can't even get the data without passing the `token`, and then `getData` always calls `checkAccess`.

keving17 replied:
gotcha. I added in both of your suggestions! Will look into fixing all of this temporary stuff after Friday!

kevinrobinson commented on server/database.js:
```diff
@@ -0,0 +1,65 @@
+    FROM evidence
+    WHERE 1=1
+      AND json->'GLOBAL'->>'location' = $1
+      AND json->>'email' IN (SELECT email FROM consented_email WHERE audio='t' AND permission='t' AND consent='t')
```
nice nested query!

keving17 commented:
Alright. I think everything outside of the s3 stuff is finished or has been pushed off until after Friday. 

I still have on ongoing discussion with @kevinrobinson about deploying and running my scripts on the production database

Let me know if you think I forgot anything.

kevinrobinson commented on scripts/prepConsent.js:
```diff
@@ -0,0 +1,67 @@
+if (process.argv.length !== 3) {
+  console.log('Command line argument missing, aborting...');
+  console.log('');
+  console.log('Usage: node tools/consent-teacher-moments.js tmp/PRIVATE_SENSITIVE_DATA/consent-teacher-moments/UserConsent.csv');
```
Script looks good!  This bit looks stale, or you could just remove this and put it in the README.

keving17 replied:
Oh right. I updated just a small part of it and forgot to check the rest of the command.

I'm planning on having this bit be in both the README and in this code - similar to how things were in the private repo

kevinrobinson commented on scripts/updateConsent.js:
```diff
@@ -0,0 +1,29 @@
+function updateConsent(database){
+  //Still need to grab emails from a sensitive data file
+  const emails = JSON.parse(fs.readFileSync('./tmp/consented-latest.json'))
+  const emailString = emails.consented.map(x => "'" + x + "'").toString();
```
⚠️ Anytime you're making SQL with strings, you're allowing SQL injection. ⚠️ 

In general, escaping is really hard to do right, and you should always use built-in functions to do this in SQL or any other language.  Check out https://stackoverflow.com/a/2049510 to see all of the craziness involved in something like email addresses where it might seem relatively safe to just concatenate, but there are a million edge cases that will work around just adding `'` around the value.  Escaping strings yourself is really hard to do correctly and securely.

I don't think `unnest` isn't doing what you want it to do here anyway.  It's taking a single string value `'kevin@mit.edu', 'tim@mit.edu'` and inserting that as the `email` column for a single row, like this:

```
 id |             email             | audio | permission | consent
----+-------------------------------+-------+------------+---------
  1 | 'foo@bar.com','kevin@mit.edu' | t     | t          | t
```

Googling "node postgres insert multiple values" leads to some suggestions here: https://github.com/brianc/node-postgres/issues/957.  You could start with mapping over the values to running a separate query for each, and then return a `Promise.all` to wait for them.

keving17 commented:
@kevinrobinson and I worked through the changes he requested in person. Dismissing the review.

kevinrobinson commented on client/src/research/DynamicTable.js:
```diff
@@ -0,0 +1,355 @@
+    const sortDirection = SortDirection.ASC;
+    const sortedList = this._sortList({sortBy, sortDirection});
+
+    //Create a dictionary mapping audioID to an audio element with the audio loaded in
```
There is a lot happening in the constructor for this component :)

Is there a way to encapsulate the work to load the audio and the transcript in a separate React component?  The point of React is so you don't have to manually manage these kinds of DOM interactions, which get increasingly difficult as you add new features or modify.  In general, if there's a ton going on in the constructor or if you're ever calling `forceUpdate` it usually means you're fighting the framework.

It sounds like one way to factor that would be to have the audio cell in the table use a different render method, that renders `<AudioWithTranscript audioId={audioId} />`.  And then that component could handle fetching the audio and transcript individually for each component as part of its `componentDidMount`.  Then you wouldn't have to manage all these changes to elements and be calling `forceUpdate` and could remove all that code.  And this will make the UX better too - it will incrementally render each component as the transcript loads rather than blocking the whole UI until all the loading is done with the `Promise.all`.

So this would be an awesome thing to do in a refactor phase.

keving17 replied:
Hmm. This makes sense. I know a partially decided to do it this way because I thought it would be annoying to have the table constantly updating with new information as audio and trascripts came in. I figured updating the table twice, once when the audio came in and once when the transcripts came in would be a smoother experience.

I'm trying to remember if there were reasons from the engineering side of things. I'll make an issue so I remember to revisit this!

keving17 replied:
Issue https://github.com/mit-teaching-systems-lab/threeflows/issues/362

Just got through the rest of your comments about this. I definitely need to fix this to use separate react components.
